reactive-banana/src/Reactive/Banana/Prim/Graph.hs

```diff
@@ -47,9 +65,9 @@ getParents gr x = maybe [] id . Map.lookup x . parents $ gr
 listParents :: forall a. (Eq a, Hashable a) => Graph a -> [a]
 listParents gr = list
     where
-    -- all nodes without children
+    -- all nodes without parents
     ancestors :: [a]
-    ancestors = [x | x <- Set.toList $ nodes gr, null (getParents gr x)]
+    ancestors = Map.keys (children gr `Map.difference` parents gr)
```

Comment:
The minor optimization: 

- Since we only want to identify nodes that have no parents, we can start filtering from `children`, rather than all `nodes`, because a node cannot exist in the graph without at least one outgoing or incoming edge.
- We can simply subtract from all `children` the nodes that have at least one incoming edge, which are the keys of `parents`, since no values of `parents` are an empty list.

Reply:
Good observation about the `children`!

However, it turns out that using `Data.HashMap.difference` has no special performance benefits — the implementation just builds a new HashMap by looking up the to-be-removed elements. (Difference of `Data.Set.Set` would be a different story). I recommend sticking to the simple list filter.

I have added a comment explaining the optimization, and introduced a new auxiliary function to make the algorithm easier to read:

```suggestion
    -- We can filter from `children`, because a node without incoming edges can only be in the graph if it has outgoing edges.
    ancestors    = [x | x <- Map.keys (children gr), not (hasParents x)]
    hasParents x = Map.member x (parents gr)
```